Frameworks/Foundation/NSDecimal.mm

```diff
@@ -0,0 +1,647 @@
+#import <algorithm>
+
+#define IS_NEGATIVE(decimal) (decimal)->_isNegative
+
+#define IS_POSITIVE(decimal) !IS_NEGATIVE(decimal)
+
+#define IS_ZERO(decimal) (((decimal)->_length == 0) && IS_POSITIVE(decimal))
+
+#define INVALIDATE_DECIMAL(decimal) \
+    \
+decimal->_length = 0;               \
+    \
+decimal->_isNegative = YES;
+
+#define SET_TO_ZERO(decimal) \
+    \
+decimal->_length = 0;        \
+    \
+decimal->_isNegative = NO;   \
+    \
+NSDecimalCompact(decimal);
+
+#define VALIDATE_ARITHMETIC_INPUT(result, leftOperand, rightOperand)                 \
+    if (NSDecimalIsNotANumber(leftOperand) || NSDecimalIsNotANumber(rightOperand)) { \
+        INVALIDATE_DECIMAL(result);                                                  \
+        return NSCalculationOverflow;                                                \
+    }
+
+const int _MAX_EXP = 127;
+
+// Internal struct to manipulate the NSDecimal struct
+// this takes care of carry over/remainder
+typedef struct {
+    NSDecimal decimal;
+    // Pointer to the mantissa being manipulated
+    unsigned short* _mantissa;
+    unsigned short _remainder;
+    // length of the array being pointed by _mantissa
+    unsigned int _extendedLen;
+} NSDecimalInternal;
+
+// Internal helper methods. All of these methods assume the NSDecimal/NSDecimalInternal is
+// normalized via NSDecimalNormalize (the exponents of the both decimals will be the same)
+static void _InternalDecimalCopy(NSDecimal* destination, NSDecimalInternal* source);
+static NSComparisonResult _compareMantissa(NSDecimal* left, NSDecimal* right);
+
+static NSCalculationError _handleDifferenceOfSignAddition(NSDecimal* result,
+                                                          NSDecimal* left,
+                                                          NSDecimal* right,
+                                                          NSRoundingMode roundingMode);
+
+static NSCalculationError _subtraction(NSDecimal* result, const NSDecimal* left, const NSDecimal* right);
+static NSCalculationError _mantissaAddition(NSDecimalInternal* result, NSDecimalInternal* augend, unsigned short addend);
+static NSCalculationError _mantissaMultiplication(NSDecimalInternal* result, NSDecimalInternal* multiplicand, unsigned short multiplier);
+static unsigned int _obtainLength(unsigned short* mantissa, int length);
+static NSCalculationError _mantissaDivision(NSDecimalInternal* result, NSDecimalInternal* dividend, unsigned short divisor);
+static NSCalculationError _compactMantissa(NSDecimalInternal* iDecimal, NSRoundingMode roundingMode);
+
+/**
+ @Status Interoperable
+*/
+void NSDecimalCopy(NSDecimal* destination, const NSDecimal* source) {
+    if (!source || !destination) {
+        return;
+    }
+
+    destination->_isNegative = source->_isNegative;
+    destination->_exponent = source->_exponent;
+    destination->_length = source->_length;
+    destination->_isCompact = source->_isCompact;
+    destination->_reserved = source->_reserved;
```

Comment:
this can be memcpy-ed and so it the mantissa. I think at point ben suggested leave it as assignment for now.


---
In reply to: [81205825](https://github.com/Microsoft/WinObjC/pull/1032#discussion_r81205825) [](ancestors = 81205825)